File: backend/app/domain/academics/schemas/course.py
"""Course timetable payloads feeding the personalized daily digest."""

from __future__ import annotations

from datetime import time
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class CourseEnrolmentInput(BaseModel):
    course_code: str = Field(min_length=2, max_length=24)
    course_title: str | None = Field(default=None, max_length=200)
    section_code: str | None = Field(default=None, max_length=24)
    term_code: str = Field(min_length=2, max_length=12)
    meeting_weekdays: list[int] = Field(default_factory=list)
    starts_at: time | None = None
    ends_at: time | None = None
    location: str | None = Field(default=None, max_length=160)
    instructor: str | None = Field(default=None, max_length=160)

    @field_validator("meeting_weekdays")
    @classmethod
    def _validate_weekdays(cls, value: list[int]) -> list[int]:
        if any(day < 1 or day > 7 for day in value):
            raise ValueError("Weekdays use ISO numbering: 1 is Monday and 7 is Sunday.")
        return sorted(set(value))


class CourseEnrolmentRecord(CourseEnrolmentInput):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    is_active: bool


class CourseScheduleEntry(BaseModel):
    """One class occurring on the digest's date, already sorted by start time."""

    course_code: str
    course_title: str | None = None
    section_code: str | None = None
    starts_at: time | None = None
    ends_at: time | None = None
    location: str | None = None

File: frontend/src/shared/api/contracts/academics/course.ts
import type {
  IsoTimeString,
  UuidString,
} from '@/shared/api/contracts/common/scalarTypes'

/**
 * `meeting_weekdays` uses ISO numbering: 1 is Monday through 7 is Sunday. The backend sorts and
 * de-duplicates the list, so the editor may submit selections in any order.
 *
 * meeting_weekdays 采用 ISO 编号：1 为周一，7 为周日。后端会排序去重，编辑器可按任意顺序提交。
 */
export interface CourseEnrolmentInput {
  course_code: string
  course_title?: string | null
  section_code?: string | null
  term_code: string
  meeting_weekdays: number[]
  starts_at?: IsoTimeString | null
  ends_at?: IsoTimeString | null
  location?: string | null
  instructor?: string | null
}

export interface CourseEnrolmentRecord extends CourseEnrolmentInput {
  id: UuidString
  is_active: boolean
}

/** One class occurring on the digest's date, already sorted by start time. */
export interface CourseScheduleEntry {
  course_code: string
  course_title?: string | null
  section_code?: string | null
  starts_at?: IsoTimeString | null
  ends_at?: IsoTimeString | null
  location?: string | null
}
